lib/livebook/completion.ex

```diff
@@ -0,0 +1,810 @@
+    if ensure_loaded?(mod) do
+      {format, docs} = get_docs(mod, [:function, :macro])
+      specs = get_specs(mod)
+      funs = funs || exports(mod)
+      funs_with_base_arity = funs_with_base_arity(docs)
+
+      funs
+      |> Enum.filter(fn {name, _arity} ->
+        name = Atom.to_string(name)
+        String.starts_with?(name, hint)
+      end)
+      |> Enum.map(fn {name, arity} ->
+        base_arity = Map.get(funs_with_base_arity, {name, arity}, arity)
+        doc = find_doc(docs, {name, base_arity})
+        spec = find_spec(specs, {name, base_arity})
+
+        docstr = doc |> doc_content(format) |> format_doc_content()
+        signatures = doc |> doc_signatures() |> format_signatures(mod)
+        spec = format_spec(spec)
+
+        %{
+          label: "#{name}/#{arity}",
+          kind: :function,
+          detail: signatures,
+          documentation: documentation_join([docstr, spec]),
+          insert_text: Atom.to_string(name)
+        }
+      end)
+    else
+      []
+    end
+  end
+
+  # If a function has default arguments it generates less-arity functions,
+  # but they have the same docs/specs as the original function.
+  # Here we build a map that given function {name, arity} returns its base arity.
+  defp funs_with_base_arity(docs) do
+    for {{_, fun_name, arity}, _, _, _, metadata} <- docs,
+        count = Map.get(metadata, :defaults, 0),
+        count > 0,
+        new_arity <- (arity - count)..(arity - 1),
+        into: %{},
+        do: {{fun_name, new_arity}, arity}
+  end
+
+  defp get_docs(mod, kinds) do
+    case Code.fetch_docs(mod) do
+      {:docs_v1, _, _, format, _, _, docs} ->
+        docs = for {{kind, _, _}, _, _, _, _} = doc <- docs, kind in kinds, do: doc
+        {format, docs}
+
+      _ ->
+        {nil, []}
+    end
+  end
+
+  defp get_module_doc_content(mod) do
+    case Code.fetch_docs(mod) do
+      {:docs_v1, _, _, format, %{"en" => docstring}, _, _} ->
+        {format, docstring}
+
+      _ ->
+        nil
+    end
+  end
+
+  defp find_doc(docs, {name, arity}) do
+    Enum.find(docs, &match?({{_, ^name, ^arity}, _, _, _, _}, &1))
+  end
+
+  defp get_specs(mod) do
+    case Code.Typespec.fetch_specs(mod) do
+      {:ok, specs} -> specs
+      :error -> []
+    end
+  end
+
+  defp find_spec(specs, {name, arity}) do
+    Enum.find(specs, &match?({{^name, ^arity}, _}, &1))
+  end
+
+  defp doc_signatures({_, _, signatures, _, _}), do: signatures
+  defp doc_signatures(_), do: []
+
+  defp doc_content({_, _, _, %{"en" => docstr}, _}, format), do: {format, docstr}
+  defp doc_content(_doc, _format), do: nil
+
+  defp documentation_join(list) do
+    case Enum.reject(list, &is_nil/1) do
+      [] -> nil
+      parts -> Enum.join(parts, "\n\n")
+    end
+  end
+
+  defp format_signatures([], _mod), do: nil
+
+  defp format_signatures(signatures, mod) do
+    prefix = mod_to_prefix(mod)
+    Enum.map_join(signatures, "\n", &(prefix <> &1))
```

Comment:
Unfortunately we cannot assume that each entry in the signature is going to be a function clause. For example, the example given by EEP 48 is this `[<<"binary_to_atom(Binary, Encoding)">>, <<"when is_binary(Binary)">>]`. YOu probably want this:

```suggestion
  defp format_signatures(signatures, mod) do
    mod_to_prefix(mod) <> Enum.join(signatures, "\n")
```